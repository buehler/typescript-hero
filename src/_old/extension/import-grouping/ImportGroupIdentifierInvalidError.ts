/**
 * Thrown when a import group identifier neither does match a keyword nor a regex pattern.
 * 
 * @export
 * @class ImportGroupIdentifierInvalidError
 * @extends {Error}
 */
export class ImportGroupIdentifierInvalidError extends Error {
    constructor(identifier: string) {
        super();
        this.message = `The identifier "${identifier}" does not match a keyword or a regex pattern (/ .. /).`;
    }
}
