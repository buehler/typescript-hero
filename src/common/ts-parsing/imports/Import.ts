import { Generatable } from '../ts-generation';
import { Clonable } from '../utilities';
import { Node } from './Node';

/**
 * Basic import interface. Defines an import in a document.
 * If no start and end points are given, the import is considered "new".
 * 
 * @export
 * @interface Import
 */
export interface Import extends Clonable, Node, Generatable {
    /**
     * Indicates if the given import is "new" or a parsed one.
     * 
     * @type {boolean}
     * @memberOf Import
     */
    readonly isNew: boolean;

    /**
     * The library name of the given import (This actually is the part after the import statement).
     *
     * @example "express"
     *
     * @type {string}
     * @memberOf Import
     */
    libraryName: string;
}

/**
 * Basic interface for aliased imports. Defines an alias for namespace imports and other aliased imports.
 * 
 * @export
 * @interface AliasedImport
 * @extends {Import}
 */
export interface AliasedImport extends Import {
    /**
     * Alias for the given import. E.g. for a "* as ..." import.
     * 
     * @type {string}
     * @memberOf AliasedImport
     */
    alias: string;
}
