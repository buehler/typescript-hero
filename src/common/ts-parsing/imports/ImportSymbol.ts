import { Generatable } from '../../ts-generation';
import { Clonable } from '../../utilities';

/**
 * Class for symbols that are contained in a named import.
 * (i.e. import {SYMBOL} from '...').
 * 
 * @export
 * @class ImportSymbol
 * @implements {Clonable}
 */
export class ImportSymbol implements Clonable, Generatable {
    constructor(public specifier: string, public alias?: string) { }

    /**
     * Generates typescript code out of the actual import.
     * 
     * @returns {string}
     * 
     * @memberOf DefaultImport
     */
    public generateTypescript(): string {
        return `${this.specifier}${this.alias ? ` as ${this.alias}` : ''}`;
    }

    /**
     * Clones the current resolve specifier and returns a new instance with the same properties.
     * 
     * @returns {ImportSymbol}
     * 
     * @memberOf ImportSymbol
     */
    public clone(): ImportSymbol {
        return new ImportSymbol(this.specifier, this.alias);
    }
}
