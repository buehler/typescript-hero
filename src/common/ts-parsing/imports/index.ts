export * from './DefaultImport';
export * from './ExternalModuleImport';
export * from './Import';
export * from './NamedImport';
export * from './NamespaceImport';
export * from './StringImport';
