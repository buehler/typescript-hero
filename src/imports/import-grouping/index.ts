export * from './import-group';
export * from './import-group-identifier-invalid-error';
export * from './import-group-keyword';
export * from './import-group-order';
export * from './import-group-setting-parser';
export * from './keyword-import-group';
export * from './regex-import-group';
export * from './remain-import-group';
