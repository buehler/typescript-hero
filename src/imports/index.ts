export { ImportAppender } from './import-appender';
export { ImportOrganizer } from './import-organizer';
