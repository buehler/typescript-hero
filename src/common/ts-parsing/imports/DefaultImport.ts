import { GenerationOptions } from '../../ts-generation';
import { AliasedImport, importRange } from './Import';
import { Serializable } from 'ts-json-serializer';
import { Range, TextDocument } from 'vscode-languageserver-types';

/**
 * Default import. Imports the default exports of a file.
 * (i.e. import foobar from ...).
 * 
 * @export
 * @class DefaultImport
 * @implements {AliasedImport}
 */
@Serializable({ factory: json => new DefaultImport(json.libraryName, json.alias, json.start, json.end) })
export class DefaultImport implements AliasedImport {
    public get isNew(): boolean {
        return this.start !== undefined && this.end !== undefined;
    }

    constructor(public libraryName: string, public alias: string, public start?: number, public end?: number) { }

    /**
     * Generates typescript code out of the actual import.
     * 
     * @param {GenerationOptions} {stringQuoteStyle, eol}
     * @returns {string}
     * 
     * @memberOf DefaultImport
     */
    public generateTypescript({stringQuoteStyle, eol}: GenerationOptions): string {
        return `import ${this.alias} from ${stringQuoteStyle}${this.libraryName}${stringQuoteStyle}${eol}\n`;
    }

    /**
     * Calculates the document range of the node in the given document.
     * 
     * @param {TextDocument} document
     * @returns {Range}
     * 
     * @memberOf DefaultImport
     */
    public getRange(document: TextDocument): Range {
        return importRange(document, this.start, this.end);
    }

    /**
     * Clone the current import object.
     * 
     * @returns {DefaultImport}
     * 
     * @memberOf DefaultImport
     */
    public clone(): DefaultImport {
        return new DefaultImport(this.libraryName, this.alias, this.start, this.end);
    }
}
