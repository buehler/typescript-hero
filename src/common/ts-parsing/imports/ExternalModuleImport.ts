import { GenerationOptions } from '../../ts-generation';
import { DocumentLike } from '../Node';
import { AliasedImport, importRange } from './Import';
import { Serializable } from 'ts-json-serializer';
import { Range } from 'vscode-languageserver-types';

/**
 * Alternative to the namespace import. Can be used by various libraries.
 * (i.e. import foobar = require('foobar')).
 * 
 * @export
 * @class ExternalModuleImport
 * @implements {AliasedImport}
 */
@Serializable({ factory: json => new ExternalModuleImport(json.libraryName, json.alias, json.start, json.end) })
export class ExternalModuleImport implements AliasedImport {
    public get isNew(): boolean {
        return this.start !== undefined && this.end !== undefined;
    }

    constructor(public libraryName: string, public alias: string, public start?: number, public end?: number) { }

    /**
     * Generates typescript code out of the actual import.
     * 
     * @param {GenerationOptions} {stringQuoteStyle, eol}
     * @returns {string}
     * 
     * @memberOf ExternalModuleImport
     */
    public generateTypescript({stringQuoteStyle, eol}: GenerationOptions): string {
        return `import ${this.alias} = require(${stringQuoteStyle}${this.libraryName}${stringQuoteStyle})${eol}\n`;
    }

    /**
     * Calculates the document range of the node in the given document.
     * 
     * @param {DocumentLike} document
     * @returns {Range}
     * 
     * @memberOf ExternalModuleImport
     */
    public getRange(document: DocumentLike): Range {
        return importRange(document, this.start, this.end);
    }

    /**
     * Clone the current import object.
     * 
     * @returns {ExternalModuleImport}
     * 
     * @memberOf ExternalModuleImport
     */
    public clone(): ExternalModuleImport {
        return new ExternalModuleImport(this.libraryName, this.alias, this.start, this.end);
    }
}
