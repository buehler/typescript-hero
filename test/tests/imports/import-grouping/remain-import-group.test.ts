import { join } from 'path';
import {
  File,
  Generatable,
  GENERATORS,
  TypescriptCodeGenerator,
  TypescriptGenerationOptions,
  TypescriptParser,
} from 'typescript-parser';
import { Uri, workspace } from 'vscode';

import { RemainImportGroup } from '../../../../src/imports/import-grouping';
import ioc from '../../../../src/ioc';
import iocSymbols, { TypescriptCodeGeneratorFactory } from '../../../../src/ioc-symbols';
import { expect } from '../../setup';

describe('RemainImportGroup', () => {

  const rootPath = workspace.workspaceFolders![0].uri.fsPath;
  const fsFile = Uri.file(
    join(rootPath, 'imports', 'import-grouping', 'imports.ts'),
  );
  let file: File;
  let importGroup: RemainImportGroup;
  let generator: TypescriptCodeGenerator;

  before(() => {
    if (!GENERATORS[RemainImportGroup.name]) {
      GENERATORS[RemainImportGroup.name] = (generatable: Generatable, options: TypescriptGenerationOptions): string => {
        const gen = new TypescriptCodeGenerator(options);
        const group = generatable as RemainImportGroup;
        if (!group.imports.length) {
          return '';
        }
        return group.sortedImports
          .map(imp => gen.generate(imp))
          .join('\n') + '\n';
      };
    }
  });

  before(async () => {
    const parser = ioc.get<TypescriptParser>(iocSymbols.parser);
    generator = ioc.get<TypescriptCodeGeneratorFactory>(iocSymbols.generatorFactory)(fsFile);
    file = await parser.parseFile(fsFile.fsPath, rootPath);
  });

  beforeEach(() => {
    importGroup = new RemainImportGroup();
  });

  it('should process all imports', () => {
    expect(file.imports.map(i => importGroup.processImport(i))).to.matchSnapshot();
  });

  it('should generate the correct typescript (asc)', () => {
    for (const imp of file.imports) {
      if (importGroup.processImport(imp)) {
        continue;
      }
    }
    expect(generator.generate(importGroup as any)).to.matchSnapshot();
  });

  it('should generate the correct typescript (desc)', () => {
    (importGroup as any).order = 'desc';
    for (const imp of file.imports) {
      if (importGroup.processImport(imp)) {
        continue;
      }
    }
    expect(generator.generate(importGroup as any)).to.matchSnapshot();
  });

});
