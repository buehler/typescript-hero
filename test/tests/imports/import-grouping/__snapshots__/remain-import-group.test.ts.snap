// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemainImportGroup should generate the correct typescript (asc) 1`] = `
"import './workspaceSideEffectLib';
import 'sideEffectLib';
import { AnotherFoobar } from './anotherFile';
import { Foobar } from './myFile';
import { AnotherModuleFoo } from 'anotherLib';
import { ModuleFoobar } from 'myLib';
"
`;

exports[`RemainImportGroup should generate the correct typescript (desc) 1`] = `
"import 'sideEffectLib';
import './workspaceSideEffectLib';
import { ModuleFoobar } from 'myLib';
import { AnotherModuleFoo } from 'anotherLib';
import { Foobar } from './myFile';
import { AnotherFoobar } from './anotherFile';
"
`;

exports[`RemainImportGroup should process all imports 1`] = `
Array [
  true,
  true,
  true,
  true,
  true,
  true,
]
`;
